### ISSUE TITLE
Reducing protobuf dependencies?

### ISSUE THREAD

**Issue by thaidn** [cgrushko/proto_library#4]
Hi cgrushko@,

First off thanks for bring protobuf support to Bazel.

We have 5 protobuf dependencies in our WORKSPACE: https://github.com/google/tink/blob/master/WORKSPACE#L43-L85.

It seems that Bazel has to download and compile each package separately. This takes a lot of time, especially compiling. Since these protobuf packages have shared code which I suspect that got compiled several times.

Is there a way to merge these packages to a single dependency so that everything is compiled only once?

Thanks!

**Comment by cgrushko** [comment]
There's some discussion about this in https://github.com/bazelbuild/bazel/issues/3219 and @jart also had some ideas how to fix this.

Right now the only option is to override the name of the external repository that proto rules depend on. You can obviously only coalesce repos that are the same (for example, JavaLite is a separate branch from Java, so you can't use one repo for both of them)

```
blaze build --proto_toolchain_for_java="@com_google_protobuf//:java_toolchain" --proto_toolchain_for_cc="@com_google_protobuf//:cc_toolchain"
```

and now you only have to define `com_google_protobuf`.

@xfxyjwf, does it make sense to change the defaults of these flags to all point at the same repository? it means that users will have to use Java, C++ and protoc from the same protobuf commit (and branch). It seems reasonable to me now, even if less purist.

**Comment by jart** [comment]
Yeah I haven't been able to understand yet why this designed was chosen, although I did offer quite a bit of feedback some months back when it happened. I'm not sure if it's something they would change.

One thing I will recommend, while I'm here, is using mirror URLs, so you can take advantage of go/bazel-downloader. Also I strongly recommend against using git_repository. Oh and you may find go/bazel-maven-config-generator to your liking. The java_import_external rule has also been merged into Bazel now. It's great. Example usage [here](https://github.com/bazelbuild/rules_closure/blob/master/closure/repositories.bzl) and [here](https://github.com/google/nomulus/blob/master/java/google/registry/repositories.bzl).

**Comment by xfxyjwf** [comment]
@cgrushko Making --proto_toolchain_for_java/cc point at com_google_protobuf sounds good to me. It might actually be required because protobuf repo now has a `workspace(name = "com_google_protobuf")` declaration in its WORKSPACE file and importing it as com_google_protobuf_java/cc is triggering a warning in bazel (see: https://github.com/google/protobuf/issues/3766).

**Comment by cgrushko** [comment]
Fix submitted, should surface in GitHub sometime tomorrow.

**Comment by thaidn** [comment]
This is great. Can we do the same thing for objectivec?

**Comment by cgrushko** [comment]
@sergiocampama
